### PR TITLE
feat:[CI-8922]: secure tunnel 

### PR DESCRIPTION
**Pull request by HenAmar**
## Describe your changes
https://harness.atlassian.net/wiki/spaces/~619256a6977c5b0072368ca7/pages/21507309591/ADR+Connecting+Harness+Cloud+to+private+network
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55748)

### CODE REVIEW

**Comment by raghavharness** [comment]
trigger uts

**Comment by raghavharness** [comment]
trigger commonchecks, codeformat, checkstyle, pmd, codebasehash

**Comment by raghavharness** [comment]
trigger uts

**Comment by raghavharness** [comment]
trigger uts

**Comment by raghavharness** [comment]
trigger uts

File: 120-ng-manager/src/main/java/io/harness/ng/tunnel/entities/Tunnel.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+public class Tunnel implements UuidAware, PersistentEntity {
+  @Id @dev.morphia.annotations.Id protected String uuid;
+  @NotEmpty @Trimmed protected String accountIdentifier;
+  @NotEmpty protected String port;
+  @CreatedDate Long createdAt;
+  @LastModifiedDate Long lastModifiedAt;
+
+  public static List<MongoIndex> mongoIndexes() {
+    return ImmutableList.<MongoIndex>builder()
+        .add(CompoundMongoIndex.builder()
```
Comment (harshjain12) [inline 120-ng-manager/src/main/java/io/harness/ng/tunnel/entities/Tunnel.java:51]: It is just accountId, Why are we making it compound indes ?

File: 120-ng-manager/src/main/java/io/harness/ng/tunnel/resources/TunnelResource.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  @ApiOperation(value = "Register the tunnel", nickname = "registerTunnel")
+  @Operation(operationId = "registerTunnel", summary = "Register the tunnel with given url and port",
+      responses =
+      {
+        @io.swagger.v3.oas.annotations.responses.
+        ApiResponse(description = "Returns true if tunnel registration is successful")
+      })
+  @Hidden
+  public ResponseDTO<Boolean>
+  registerTunnel(@Parameter(description = NGCommonEntityConstants.ACCOUNT_PARAM_MESSAGE) @QueryParam(
```
Comment (harshjain12) [inline 120-ng-manager/src/main/java/io/harness/ng/tunnel/resources/TunnelResource.java:89]: Can we check ACL annotations need to be pull.
Seems to be missing

File: 980-commons/src/main/java/io/harness/network/Http.java [hunk 3]
```diff
@@ -129,6 +129,41 @@ public Integer load(String url) throws IOException {
             }
           });
 
+  LoadingCache<HttpURLWithProxy, Integer> responseCodeForValidationWithProxy =
+      CacheBuilder.newBuilder()
+          .maximumSize(1000)
+          .expireAfterWrite(1, TimeUnit.MINUTES)
```
Comment (harshjain12) [inline 980-commons/src/main/java/io/harness/network/Http.java:135]: Do we need it every 1 min ?

**Review by harshjain12 — CHANGES_REQUESTED** [review]

File: 120-ng-manager/src/main/java/io/harness/ng/tunnel/entities/Tunnel.java (hunk 1 above)
Comment (raghavharness) [reply 120-ng-manager/src/main/java/io/harness/ng/tunnel/entities/Tunnel.java:51]: We want to add an index on accountId but there's no other type of index we see in code. Is there any index type which we can use here for a single field ?

File: 120-ng-manager/src/main/java/io/harness/ng/tunnel/resources/TunnelResource.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    content =
+    {
+      @Content(mediaType = NGCommonEntityConstants.APPLICATION_JSON_MEDIA_TYPE,
+          schema = @Schema(implementation = ErrorDTO.class))
+      ,
+          @Content(mediaType = NGCommonEntityConstants.APPLICATION_YAML_MEDIA_TYPE,
+              schema = @Schema(implementation = ErrorDTO.class))
+    })
+@Slf4j
+public class TunnelResource {
```
[the remark this answers is not in the document]
Comment (raghavharness) [reply 120-ng-manager/src/main/java/io/harness/ng/tunnel/resources/TunnelResource.java:76]: I saw in ng manager we are writing all the resource class like this only that's why didn't use different way. Do you think we should add it ?

File: 120-ng-manager/src/main/java/io/harness/ng/tunnel/resources/TunnelResource.java (hunk 2 above)
Comment (raghavharness) [reply 120-ng-manager/src/main/java/io/harness/ng/tunnel/resources/TunnelResource.java:89]: Have added `@NextGenManagerAuth` on top of the class.

File: 980-commons/src/main/java/io/harness/network/Http.java [hunk 5]
```diff
@@ -129,6 +129,41 @@ public Integer load(String url) throws IOException {
...
+              // Install the all-trusting host verifier
+              HttpsURLConnection.setDefaultHostnameVerifier(allHostsValid);
+              HttpURLConnection connection = getHttpsURLConnectionWithProxy(
+                  httpURLWithProxy.getUrl(), httpURLWithProxy.getProxyHost(), httpURLWithProxy.getProxyPort());
+              try {
+                // Changed to GET as some providers like artifactory SAAS is not
+                // accepting HEAD requests
+                connection.setRequestMethod("GET");
+                connection.setConnectTimeout(15000);
+                connection.setReadTimeout(15000);
```
[the remark this answers is not in the document]
Comment (raghavharness) [reply 980-commons/src/main/java/io/harness/network/Http.java:155]: yes its in ms. BTW we are not using any custom logic of ours. We just created a copy of `responseCodeForValidation` which is already used but we created one which supports proxy.

File: 980-commons/src/main/java/io/harness/network/Http.java [hunk 6]
```diff
@@ -129,6 +129,41 @@ public Integer load(String url) throws IOException {
...
           });
 
+  LoadingCache<HttpURLWithProxy, Integer> responseCodeForValidationWithProxy =
+      CacheBuilder.newBuilder()
+          .maximumSize(1000)
+          .expireAfterWrite(1, TimeUnit.MINUTES)
+          .build(new CacheLoader<HttpURLWithProxy, Integer>() {
+            @Override
+            public Integer load(HttpURLWithProxy httpURLWithProxy) throws IOException {
+              log.info("Testing connectivity");
```
[the remark this answers is not in the document]
Comment (raghavharness) [reply 980-commons/src/main/java/io/harness/network/Http.java:139]: Ditto, kept it same.

File: 980-commons/src/main/java/io/harness/network/Http.java (hunk 3 above)
Comment (raghavharness) [reply 980-commons/src/main/java/io/harness/network/Http.java:135]: Ditto, kept it same.

**Comment by raghavharness** [comment]
trigger uts

**Comment by raghavharness** [comment]
trigger uts

**Comment by raghavharness** [comment]
trigger codeformat, checkstyle, pmd, codebasehash

**Comment by raghavharness** [comment]
trigger uts

**Review by harshjain12 — APPROVED** [review]

File: 120-ng-manager/src/main/java/io/harness/ng/tunnel/entities/Tunnel.java (hunk 1 above)
Comment (harshjain12) [reply 120-ng-manager/src/main/java/io/harness/ng/tunnel/entities/Tunnel.java:51]: @FdIndex private String accountId;

Comment (raghavharness) [reply 120-ng-manager/src/main/java/io/harness/ng/tunnel/entities/Tunnel.java]: fixed

**Comment by raghavharness** [comment]
trigger uts

**Review by adivishy1 — APPROVED** [review]
> Reviewed 6 of 45 files at r1, 8 of 16 files at r4, 11 of 23 files at r6, 1 of 3 files at r7, 15 of 15 files at r8, 4 of 5 files at r9, 27 of 27 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55748)* status: 6. unresolved discussions.

**Comment by raghavharness** [comment]
trigger ut5, ut7

**Comment by raghavharness** [comment]
trigger ut7

**Review by harshjain12 — APPROVED** [review]

**Pull request merged**